File: common/src/main/java/dev/obscuria/fragmentum/api/v1/server/FragmentumServerEvents.java
package dev.obscuria.fragmentum.api.v1.server;

import dev.obscuria.fragmentum.api.v1.common.event.Event;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.server.packs.resources.CloseableResourceManager;

public interface FragmentumServerEvents
{
    Event<Start> SERVER_STARTING = Event.create();
    Event<Start> SERVER_STARTED = Event.create();

    Event<StartDataReload> START_DATA_PACK_RELOAD = Event.create();
    Event<EndDataReload> END_DATA_PACK_RELOAD = Event.create();
    Event<SyncData> SYNC_DATA_PACK_CONTENTS = Event.create();

    Event<Save> BEFORE_SAVE = Event.create();
    Event<Save> AFTER_SAVE = Event.create();

    Event<Stop> SERVER_STOPPING = Event.create();
    Event<Stop> SERVER_STOPPED = Event.create();

    @FunctionalInterface
    interface Start
    {
        void invoke(MinecraftServer server);
    }

    @FunctionalInterface
    interface StartDataReload
    {
        void invoke(MinecraftServer server, CloseableResourceManager manager);
    }

    @FunctionalInterface
    interface EndDataReload
    {
        void invoke(MinecraftServer server, CloseableResourceManager manager, boolean success);
    }

    @FunctionalInterface
    interface SyncData
    {
        void invoke(ServerPlayer player, boolean joined);
    }

    @FunctionalInterface
    interface Save
    {
        void invoke(MinecraftServer server, boolean flush, boolean force);
    }

    @FunctionalInterface
    interface Stop
    {
        void invoke(MinecraftServer server);
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/api/server/FragmentumServerEvents.kt
package dev.obscuria.fragmentum.api.server

import dev.obscuria.fragmentum.api.tools.event.Event
import net.minecraft.server.MinecraftServer
import net.minecraft.server.level.ServerPlayer
import net.minecraft.server.packs.resources.CloseableResourceManager

object FragmentumServerEvents
{
    val SERVER_STARTING: Event<Start> = Event.create()
    val SERVER_STARTED: Event<Start> = Event.create()

    val START_DATA_PACK_RELOAD: Event<StartDataReload> = Event.create()
    val END_DATA_PACK_RELOAD: Event<EndDataReload> = Event.create()
    val SYNC_DATA_PACK_CONTENTS: Event<SyncData> = Event.create()

    val BEFORE_SAVE: Event<Save> = Event.create()
    val AFTER_SAVE: Event<Save> = Event.create()

    val SERVER_STOPPING: Event<Stop> = Event.create()
    val SERVER_STOPPED: Event<Stop> = Event.create()

    fun interface Start
    {
        fun invoke(server: MinecraftServer)
    }

    fun interface StartDataReload
    {
        fun invoke(server: MinecraftServer, manager: CloseableResourceManager)
    }

    fun interface EndDataReload
    {
        fun invoke(server: MinecraftServer, manager: CloseableResourceManager, success: Boolean)
    }

    fun interface SyncData
    {
        fun invoke(player: ServerPlayer, joined: Boolean)
    }

    fun interface Save
    {
        fun invoke(server: MinecraftServer, flush: Boolean, force: Boolean)
    }

    fun interface Stop
    {
        fun invoke(server: MinecraftServer)
    }
}